pkg/contexts/ocm/blobhandler/handlers/generic/mvn/blobhandler.go

```diff
@@ -0,0 +1,190 @@
+package mvn
+
+import (
+	"context"
+	"crypto"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/mandelsoft/vfs/pkg/vfs"
+
+	"github.com/open-component-model/ocm/pkg/contexts/credentials/builtin/mvn/identity"
+	"github.com/open-component-model/ocm/pkg/contexts/ocm/accessmethods/mvn"
+	"github.com/open-component-model/ocm/pkg/contexts/ocm/cpi"
+	"github.com/open-component-model/ocm/pkg/contexts/ocm/resourcetypes"
+	"github.com/open-component-model/ocm/pkg/logging"
+	"github.com/open-component-model/ocm/pkg/mime"
+	"github.com/open-component-model/ocm/pkg/utils/tarutils"
+)
+
+const BLOB_HANDLER_NAME = "ocm/" + resourcetypes.MVN_ARTIFACT
+
+type artifactHandler struct {
+	spec *Config
+}
+
+func NewArtifactHandler(repospec *Config) cpi.BlobHandler {
+	return &artifactHandler{repospec}
+}
+
+var log = logging.Context().Logger(identity.REALM)
+
+func (b *artifactHandler) StoreBlob(blob cpi.BlobAccess, resourceType string, hint string, _ cpi.AccessSpec, ctx cpi.StorageContext) (cpi.AccessSpec, error) {
+	// check conditions
+	if b.spec == nil {
+		return nil, nil
+	}
+	mimeType := blob.MimeType()
+	if resourcetypes.MVN_ARTIFACT != resourceType {
+		log.Debug("not a MVN artifact", "resourceType", resourceType)
+		return nil, nil
+	}
+	if mime.MIME_TGZ != mimeType {
+		log.Debug("not a tarball, can't be a complete mvn GAV", "mimeType", mimeType)
+		return nil, nil
+	}
+	if b.spec.Url == "" {
+		return nil, fmt.Errorf("MVN repository url not provided")
+	}
+
+	// setup logger
+	log = log.WithValues("repository", b.spec.Url)
+
+	// identify artifact
+	artifact := mvn.ArtifactFromHint(hint)
+	log = log.WithValues("groupId", artifact.GroupId, "artifactId", artifact.ArtifactId, "version", artifact.Version)
+	log.Debug("identified")
+
+	// get credentials
+	cred := identity.GetCredentials(ctx.GetContext(), b.spec.Url, artifact.GroupPath())
+	if cred == nil {
+		return nil, fmt.Errorf("no credentials found for %s. Couldn't upload '%s'", b.spec.Url, artifact)
+	}
+	username := cred[identity.ATTR_USERNAME]
+	password := cred[identity.ATTR_PASSWORD]
+	if username == "" || password == "" {
+		return nil, fmt.Errorf("credentials for %s are invalid. Username or password missing! Couldn't upload '%s'", b.spec.Url, artifact)
+	}
+	log = log.WithValues("user", username)
+	log.Debug("found credentials")
+
+	// Create a new request
+	blobReader, err := blob.Reader()
+	if err != nil {
+		return nil, err
+	}
+	defer blobReader.Close()
+
+	tempFs, err := tarutils.ExtractTgzToTempFs(blobReader)
+	if err != nil {
+		return nil, err
+	}
+	defer vfs.Cleanup(tempFs)
+	files, err := tarutils.ListSortedFilesInDir(tempFs, "", false)
+	if err != nil {
+		return nil, err
+	}
+	for _, file := range files {
```

Comment:
you could stream the tar without unpacking it into a fs.
Just loop over the file header and directly push from the tar stream.
the hash can be calculated on the fly by the reader.

Reply:
but then the files are already deployed on the remote server and it's harder to delete them again... I prefer to upfront calc the digest... now using the new `HashReader` ;-)